bin/bundle_tools/storage.py

```diff
@@ -0,0 +1,209 @@
+
+
+class StagedBundleFinder:
+
+    def __init__(self):
+        self.bundle_paths = list()
+        self.s3 = boto3.client('s3')
+
+    def paths_of_bundles_under(self, s3url: Url) -> list:
+        # Assumption: bundles are stored at **/bundles/bundleX/
+        logger.output(f"\nFinding bundles under {str(s3url)}...")
+        self._search_for_bundles_in_folder(bucket=s3url.host, root_path=s3url.path.lstrip('/'))
+        logger.output("\n")
+        return self.bundle_paths
+
+    def _search_for_bundles_in_folder(self, bucket: str, root_path: str) -> list:
+        for folder_path in self._subfolders_in_folder(bucket, root_path):
+            if self._is_bundle_home(root_path):
+                self.bundle_paths.append(folder_path.rstrip('/'))
+            else:
+                self._search_for_bundles_in_folder(bucket, folder_path)
+
+    def _subfolders_in_folder(self, bucket: str, folder_path: str):
+        paginator = self.s3.get_paginator('list_objects')
+        for page in paginator.paginate(Bucket=bucket, Prefix=folder_path, Delimiter='/'):
+            if 'CommonPrefixes' in page:
+                for obj in page['CommonPrefixes']:
+                    yield obj['Prefix']
+
+    @staticmethod
+    def _is_bundle_home(path: str) -> bool:
+        return os.path.basename(path.rstrip('/')) == 'bundles'
+
+
+class DSSAPIError(RuntimeError):
+    pass
+
+
+class DSSDriver:
+
+    FAKE_CREATOR_UID = 104
+    DEFAULT_DSS_REPLICA = 'aws'
+    BACKOFF_FACTOR = 1.618
+
+    def __init__(self, endpoint_url, report_task_ids=False):
+        self.dss_url = endpoint_url
+        self.report_task_ids = report_task_ids
+
+    def put_file(self, bundle_uuid: str, file_uuid: str, file_location: str):
+        raise NotImplemented
+
+    def put_bundle(self, bundle_uuid: str, file_info: list):
+        raise NotImplemented
+
+
+class DSSpythonDriver(DSSDriver):
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+    def put_file(self, bundle_uuid: str, file_uuid: str, file_location: str):
+        response = hca_api.put_files(uuid=file_uuid,
+                                     source_url=file_location,
+                                     creator_uid=104,
+                                     bundle_uuid=bundle_uuid,
+                                     api_url=self.dss_url
+                                     )
+        if response.status_code != 201:
+            print(f"ERROR: put_files() returned {response.status_code}: {response.text}")
+            exit(1)
+        return response.json()['version']
+
+    def put_bundle(self, bundle_uuid: str, file_info: list):
+        response = hca_api.put_bundles(bundle_uuid, self.DEFAULT_DSS_REPLICA, self.FAKE_CREATOR_UID, file_info)
+        if response.status_code != 201:
+            print(f"ERROR: put_files() returned {response.status_code}: {response.text}")
+            exit(1)
+        return response.json()['version']
```

Comment:
Consider importing the DSS API bindings on demand within the functions, to avoid a runtime dependency on hca-cli. (Re-importing a library is very cheap in Python - just a few dict lookups I believe.)